### DISCUSSION TITLE
Nested models, cannot wrap my head around adding an additional "manual" field. Please nudge me in the proper direction.

### DISCUSSION THREAD

**Discussion by teodoryantcheff** [fastapi/fastapi#11660] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class BodyTypeBase(SQLModel):
    name: str = Field(unique=True, index=True)
    label: Optional[str] = None
    vehicle_type_id: int = Field(default=None, foreign_key='vehicletype.id')

class BodyTypeRead(BodyTypeBase):
    id: int

class VehicleBase(SQLModel):
    id: Optional[int] = Field(default=None, primary_key=True)
    make: str = Field(index=True)
    model: str = Field(index=True)
    ... 

class VehicleWithBodyType(VehicleBase):
    body_type: BodyTypeRead

    url: str = Field(default=None)   << This is NOT coming from the DB

@app.get('/vehicles/{vehicle_id}', response_model=VehicleWithBodyType)
def get_vehicle(
        db: Session = Depends(get_db),
        request: Request,
        vehicle_id: int
):
    v = db.get(Vehicle, vehicle_id)
    if v:
        return v
        # ^ This works as expected _with_ the nested VehicleWithBodyType model
        

        return {**v.model_dump(), 'url': str(request.url_for('get_vehicle', vehicle_id=vehicle_id))}
        # ^ This I can make work only if the response_model does not have nested children

    raise HTTPException(status_code=404, detail="not found")
```

### Description

I have this nested model `VehicleWithBodyType`. It has the `url` field declared which I want to be dynamically adding to the response of a `get_vehicle` endpoint.

If I make `VehicleWithBodyType` _not_ have the `body_type` nested model 
```python 
class VehicleWithBodyType(VehicleBase):
    #### body_type: BodyTypeRead  << no nested children

    url: str = Field(default=None)   << This is NOT coming from the DB

@app.get('/vehicles/{vehicle_id}', response_model=VehicleWithBodyType)
def get_vehicle(
        db: Session = Depends(get_db),
        request: Request,
        vehicle_id: int
):
    v = db.get(Vehicle, vehicle_id)
    if v:
        return {**v.model_dump(), 'url': str(request.url_for('get_vehicle', vehicle_id=vehicle_id))}
```
works like a charm and the `url` is added, no problems.

But when I make `VehicleWithBodyType` _have_ the nested `body_type: BodyTypeRead` field when trying to set the `url` on `v`  and return that I do get 
```
fastapi.exceptions.ResponseValidationError: 1 validation errors:
  {'type': 'missing', 'loc': ('response', 'body_type'), 'msg': 'Field required', 
```

Nested model and _not setting_ `url` :

```python 
class VehicleWithBodyType(VehicleBase):
    body_type: BodyTypeRead

    url: str = Field(default=None)

@app.get('/vehicles/{vehicle_id}', response_model=VehicleWithBodyType)
def get_vehicle(
        db: Session = Depends(get_db),
        request: Request,
        vehicle_id: int
):
    v = db.get(Vehicle, vehicle_id)
    if v:
        return v  # do not  set v.url 
    raise HTTPException(status_code=404, detail="not found")
```
works as expected. 
But I want to have the `url` set.

So it's either working with the nested model and no `url`, or can have/set the `url` and loosing the nested model. 

Which defies the purpose...

Spend hours trying to understand what I'm doing wrong.
Read the [fastapi docs on responce_model](https://fastapi.tiangolo.com/tutorial/response-model/?h=respo#response_model-parameter) many, many times.
Read [pydantic's serializations docs](https://docs.pydantic.dev/latest/concepts/serialization/), and the [api docs](https://docs.pydantic.dev/latest/api/base_model/#pydantic.BaseModel.model_dump).
Stepped through fastapi's code (`serialize_response` in `.venv/Lib/site-packages/fastapi/routing.py`)

What am I missing?

### Operating System

Linux, Windows

### FastAPI Version

0.111.0

### Pydantic Version

2.7.2

### Python Version

Python 3.11.9

**Comment by n8sty** [comment]
How about something like:

```
@app.get('/vehicles/{vehicle_id}', response_model=VehicleWithBodyType)
def get_vehicle(
        db: Session = Depends(get_db),
        request: Request,
        vehicle_id: int
):
    v = db.get(Vehicle, vehicle_id)
    if v:
        data = v.dict()
        return {"body_type": data, "url": str(request.url_for('get_vehicle', vehicle_id=vehicle_id))}
    raise HTTPException(status_code=404, detail="not found")
```

**Reply by teodoryantcheff** [reply]
Thanks for your reply, @n8sty !

What you suggest will I think nest the `Vehicle` data coming from the database under a `body_type` key. And while it will likely solve the serialization exception it will also make the responses not be a `Vehicle` root model but something else.

Essentially I'm "just" trying to set a field value to an already populated `Vehicle` sqlmodel/pydantic model. And failing miserably at that... :)

**Comment by YuriiMotov** [comment]
Let's close this for now to keep the discussion clean.
If you’ve found a solution, please share it here.
If you still need help, feel free to let us know.
